### PR TITLE
[python] add type hints for custom objective and metric functions in scikit-learn interface

### PR DESCRIPTION
**Pull request by jameslamb**
Created in response to https://github.com/microsoft/LightGBM/pull/4544#pullrequestreview-735574138.

Contributes to #3756.

Proposes introducing more specific type hints for custom objective and metric functions in the scikit-learn and Dask interfaces in the Python package.

### CODE REVIEW

File: python-package/lightgbm/sklearn.py [hunk 1]
```diff
@@ -2,22 +2,52 @@
...
+_GroupType = Union[_ArrayLike, List[int]]
+
+_LGBM_ScikitCustomObjectiveFunction = Union[
+    Callable[
+        [_ArrayLike, _ArrayLike],
+        Tuple[np.ndarray, np.ndarray]
+    ],
+    Callable[
+        [_ArrayLike, _ArrayLike, _GroupType],
+        Tuple[np.ndarray, np.ndarray]
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply python-package/lightgbm/sklearn.py]: Sorry for taking so long to get back to this one!

I just pushed https://github.com/microsoft/LightGBM/pull/4547/commits/ea1aada69350cc53bc7ee2ac628400c7d622fdb1 with my best understanding of your comments above, but to be honest I still am confused about exactly what is allowed.

Here is my interpretation of those comments / links:

* eval function
    - `y_true` = list, numpy array, or pandas Series
    - `y_pred` = numpy array
    - `group` = numpy array
    - `weight` = numpy array
* objective function
    - `y_true` = list, numpy array, or pandas Series
    - `y_pred` = numpy array
    - `group` = numpy array
    - `grad` (output) = list, numpy array, or pandas Series
    - `hess` (output) = list, numpy array, or pandas Series

File: python-package/lightgbm/sklearn.py [hunk 2]
```diff
@@ -2,23 +2,52 @@
...
+_ArrayLike = Union[List, np.ndarray, pd_Series]
+_EvalResultType = Tuple[str, float, bool]
+
+_LGBM_ScikitCustomObjectiveFunction = Union[
+    Callable[
+        [_ArrayLike, np.ndarray],
+        Tuple[np.ndarray, np.ndarray]
+    ],
+    Callable[
+        [_ArrayLike, np.ndarray, np.ndarray],
```
Comment (shiyu1994) [inline python-package/lightgbm/sklearn.py:26]: Shall we allow the users to pass `list` or `pd_Series` to the second parameter `y_pred` and the third parameter `group`, by changing the second and third argument type to `_ArrayLike`?

Comment (shiyu1994) [reply python-package/lightgbm/sklearn.py:26]: Sorry for missing your discussion with @StrikerRUS above. After reading the comments, I agree with the current types for these parameters.

File: python-package/lightgbm/sklearn.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shiyu1994) [reply python-package/lightgbm/sklearn.py]: Double-checked this and I think your interpretation is fine. Thanks for the detailed investigation.

**Review by shiyu1994 — APPROVED** [review]

File: python-package/lightgbm/sklearn.py [hunk 3]
```diff
@@ -11,14 +11,42 @@
...
+        Union[_EvalResultType, List[_EvalResultType]]
+    ],
+    Callable[
+        [_ArrayLike, np.ndarray, np.ndarray, np.ndarray],
+        Union[_EvalResultType, List[_EvalResultType]]
+    ],
+]
```
Comment (StrikerRUS) [inline python-package/lightgbm/sklearn.py:43]: `y_true` cannot be anything else than `np.ndarray` because it is result of `get_label()` method:
https://github.com/microsoft/LightGBM/blob/144036f2df5b584b8302742c7036545f90f08fdb/python-package/lightgbm/sklearn.py#L173
https://github.com/microsoft/LightGBM/blob/144036f2df5b584b8302742c7036545f90f08fdb/python-package/lightgbm/basic.py#L2303-L2310
```suggestion
_LGBM_ScikitCustomObjectiveFunction = Union[
    Callable[
        [np.ndarray, np.ndarray],
        Tuple[_ArrayLike, _ArrayLike]
    ],
    Callable[
        [np.ndarray, np.ndarray, np.ndarray],
        Tuple[_ArrayLike, _ArrayLike]
    ],
]
_LGBM_ScikitCustomEvalFunction = Union[
    Callable[
        [np.ndarray, np.ndarray],
        Union[_EvalResultType, List[_EvalResultType]]
    ],
    Callable[
        [np.ndarray, np.ndarray, np.ndarray],
        Union[_EvalResultType, List[_EvalResultType]]
    ],
    Callable[
        [np.ndarray, np.ndarray, np.ndarray, np.ndarray],
        Union[_EvalResultType, List[_EvalResultType]]
    ],
]
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
I'm sorry, my [previous comment](https://github.com/microsoft/LightGBM/pull/4547#discussion_r696045024) was very vague. But you've got almost everything right from it! 😄

File: python-package/lightgbm/sklearn.py (hunk 3 above)
Comment (jameslamb) [reply python-package/lightgbm/sklearn.py:43]: ah ok, thank you

File: python-package/lightgbm/dask.py [hunk 4]
```diff
@@ -1096,7 +1097,7 @@ def __init__(
         learning_rate: float = 0.1,
         n_estimators: int = 100,
         subsample_for_bin: int = 200000,
-        objective: Optional[Union[Callable, str]] = None,
+        objective: Optional[Union[_LGBM_ScikitCustomObjectiveFunction, str]] = None,
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply python-package/lightgbm/dask.py]: yeah fair point! Updated in https://github.com/microsoft/LightGBM/pull/4547/commits/ccca96597d25418e74e43e854b45ab3d15aeadfa

**Review by StrikerRUS — APPROVED** [review]
Thank you very much for the deep investigation into accepted types!

**Comment by jameslamb** [comment]
Thanks again for the help @StrikerRUS , this one required a lot of investigation 😄
